### PR TITLE
HAAR-894: Add Validation of email domains

### PR DESCRIPTION
**Pull request by karenmillermoj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/jpa/repository/EmailDomainRepository.kt [hunk 1]
```diff
@@ -1,9 +1,13 @@
...
 import org.springframework.data.repository.CrudRepository
 import uk.gov.justice.digital.hmpps.externalusersapi.model.EmailDomain
 import java.util.UUID
 
 interface EmailDomainRepository : CrudRepository<EmailDomain, UUID> {
   fun findByName(name: String): EmailDomain?
+
+  @Query(value = "SELECT * FROM email_domain WHERE :domainName like name ", nativeQuery = true)
+  fun findByNameLike(domainName: String): EmailDomain?
 }
```
Comment (karenmillermoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/jpa/repository/EmailDomainRepository.kt:13]: I did this as a native query as we are trying to search on a field that contains a "%" and the jpa call kept escaping the column value. 
Note.... the like check is the opposite way round to what you would normally expect as the stored value holds a "%"
e.g. select * from email_domain where 'fred.smith.co.uk' like '%smith.co.uk'

File: src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/integration/resource/EmailDomainsIntTest.kt [hunk 2]
```diff
@@ -125,7 +125,7 @@ class EmailDomainsIntTest : IntegrationTestBase() {
         )
       )
       .exchange()
-      .expectStatus().is4xxClientError
+      .expectStatus().isBadRequest
```
Comment (karenmillermoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/integration/resource/EmailDomainsIntTest.kt:128]: more specific status

File: src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/resource/ValidationController.kt [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+@Tag(name = "/validate", description = "Validation Controller")
+class ValidationController(
+  private val verifyEmailDomainService: VerifyEmailDomainService
+) {
+  @GetMapping("/validate/email-domain")
+  @Operation(
+    summary = "Validates Email domain",
+    description = "Validates Email domain."
+  )
+  @ApiResponses(
```
[the remark this answers is not in the document]
Comment (karenmillermoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/resource/ValidationController.kt:27]: Not sure if it's good enough - but it shows:

Example value | Schema
true

The swagger docs seems to take care of it for us (like it does with any object).

File: src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/resource/ValidationControllerIntTest.kt [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        .headers(setAuthorisation("AUTH_ADM", listOf("ROLE_CREATE_EMAIL_TOKEN")))
+        .exchange()
+        .expectStatus().isOk
+        .expectBody<Boolean>()
+        .returnResult().responseBody
+
+      assertThat(isValid).isEqualTo(false)
+    }
+  }
+
```
[the remark this answers is not in the document]
Comment (karenmillermoj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/resource/ValidationControllerIntTest.kt:27]: prob not a good name
the existing email domains matches as the email domains list has: careuk.com
so both careuk.com and 1careuk.com match

The second test is a bit like a subdomain (obv without the .)
 I can rename the tests to:
Validate email matching domain and Validate email domain matching subdomain ?

[the remark this answers is not in the document]
Comment (AlTurner-MOJ) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/resource/ValidationControllerIntTest.kt:27]: Ah ok, makes sense

**Review by AlTurner-MOJ — DISMISSED** [review]
lgtm

**Review by AlTurner-MOJ — APPROVED** [review]
lgtm

**Pull request merged**